### PR TITLE
IncrementalDB for new node and seed node joining

### PR DESCRIPTION
**Pull request by sandipbhoir**
## Description:
**Pre-launch of zilliqa process:** (Not part of this PR)
 - downloadIncrDB.py will fetch the following from S3:
    - Entire persistence from S3 
    - StateDeltas from S3 ( All state deltas from last 10 DS epoch if available )

**This PR covers following:**

**Launch of zilliqa process:**
 - New node and seed node will be started with --recovery option.
 - This will enable retrieve from existing local persistence downloaded pre-launch of zilliqa where states    will be regenerated for last 10 DS epoch (or less depending on available state-deltas ).
 - After this, both will start syncronization from lookup. However, instead of asking for state from lookup, it will ask for state-deltas to generate latest state.
  

## Backward Compatibility

- [ ] This is not a breaking change
- [x] This is a breaking change

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [ ] local machine test

- [x] small-scale cloud test

### CODE REVIEW

**Comment by ckyang** [comment]
For incremental DB things, it's better to verify like that:

1. Launch a small-scale (ex. 33 nodes) testnet with current master
2. Use this PR + https://github.com/Zilliqa/testnet/pull/426 to upgrade (1) to a separated testnet
    (doc: https://docs.google.com/spreadsheets/d/1J5wVeY3o4q7SIu38HbvUdHkdV3FAyCMWgSNg8wnuOjs/edit#gid=2001092950)
3. See if anything works fine in the separated testnet

File: src/libLookup/Lookup.cpp [hunk 1]
```diff
@@ -1825,6 +1974,21 @@ void Lookup::CommitTxBlocks(const vector<TxBlock>& txBlocks) {
...
+  if (m_syncType == SyncType::NEW_SYNC ||
+      m_syncType == SyncType::NEW_LOOKUP_SYNC) {  // only for new node joining
+    // Get the state-delta for all txBlocks from random lookup nodes
+    GetStateDeltasFromSeedNodes(lowBlockNum, highBlockNum);
+
+    std::unique_lock<std::mutex> cv_lk(m_mutexSetStateDeltaFromSeed);
+    if (cv_setStateDeltasFromSeed.wait_for(
+            cv_lk, std::chrono::seconds(GETSTATEDELTAS_TIMEOUT_IN_SECONDS)) ==
+        std::cv_status::timeout) {
+      LOG_GENERAL(WARNING, "Didn't receive statedeltas!");
```
Comment (kaikawaliu) [inline src/libLookup/Lookup.cpp]: If the node missed the first time for fetching delta, will the querying in the next try includes the deltas that it failed to fetch previously?

File: src/libLookup/Lookup.cpp [hunk 2]
```diff
@@ -1901,6 +2085,69 @@ bool Lookup::ProcessSetStateDeltaFromSeed(const bytes& message,
...
+
+  bool isAnyTxBlkFromVacuousEpoch = false;
+  for (auto i = lowBlockNum; i <= highBlockNum; i++) {
+    if ((i + 1) % NUM_FINAL_BLOCK_PER_POW == 0) {
+      isAnyTxBlkFromVacuousEpoch = true;
+      break;
+    }
+  }
+
+  if (isAnyTxBlkFromVacuousEpoch &&
```
Comment (kaikawaliu) [inline src/libLookup/Lookup.cpp]: This judge is a bit weird. E.g. a node trying to sync at epoch 21 (10 tx epoch / ds), the whole states it fetched is at epoch 10 (very likely to happen as the states uploading may not finish). Then the state delta is from epoch 11 ~ 21. Which includes an vacuous epoch. Then in this case, we will result in calling `MoveUpdatesToDisk` for the state in epoch 21, this is what we don't want.
So the change I suggest is to call `MoveUpdatesToDisk` only when the highBlockNum being the vacuous block.

File: src/libMessage/Messenger.cpp [hunk 3]
```diff
@@ -5432,6 +5518,45 @@ bool Messenger::GetLookupSetStateDeltaFromSeed(const bytes& src,
...
+
+  result.ParseFromArray(src.data() + offset, src.size() - offset);
+
+  if (!result.IsInitialized()) {
+    LOG_GENERAL(WARNING, "LookupSetStateDeltasFromSeed initialization failed");
+    return false;
+  }
+
+  lowBlockNum = result.data().lowblocknum();
+  highBlockNum = result.data().highblocknum();
```
Comment (kaikawaliu) [inline src/libMessage/Messenger.cpp]: It will be better to have a check on the size of delta and the (highBlockNum - lowBlockNum) is needed

File: src/libLookup/Lookup.cpp (hunk 1 above)
Comment (sandipbhoir) [reply src/libLookup/Lookup.cpp]: Agree. Will try to keep track of missing statedeltas and Try fetching them next time.

File: src/libLookup/Lookup.cpp (hunk 2 above)
Comment (sandipbhoir) [reply src/libLookup/Lookup.cpp]: Agree.

File: src/libLookup/Lookup.cpp [hunk 4]
```diff
@@ -1901,6 +2085,69 @@ bool Lookup::ProcessSetStateDeltaFromSeed(const bytes& message,
...
+                  << lookupPubKey << " is not in my lookup node list.");
+    return false;
+  }
+
+  LOG_EPOCH(INFO, m_mediator.m_currentEpochNum,
+            "ProcessSetStateDeltasFromSeed sent by "
+                << from << " for blocks: " << lowBlockNum << " to "
+                << highBlockNum);
+
+  for (const auto& delta : stateDeltas) {
```
Comment (kaikawaliu) [inline src/libLookup/Lookup.cpp]: I think we need to verify the state delta against the state delta hash in the corresponding txblock. One way to locate the txBlock is to incrementally using the count in the vector and the lowBlockNum. Another way can be modifying the message structure to indicate the block number of each delta, which may require extra code change however but it's better for future maintenance.

File: src/libNode/Node.cpp [hunk 5]
```diff
@@ -134,6 +134,11 @@ bool Node::Install(const SyncType syncType, const bool toRetrieveHistory) {
       return false;
     }
 
+    if (SyncType::NEW_SYNC == syncType ||
+        SyncType::NEW_LOOKUP_SYNC == syncType) {
+      return true;
+    }
```
Comment (kaikawaliu) [inline src/libNode/Node.cpp]: If we don't skip what will happen?

File: src/libNode/Node.cpp [hunk 6]
```diff
@@ -1855,26 +1863,9 @@ bool Node::ProcessDSGuardNetworkInfoUpdate(const bytes& message,
   LOG_GENERAL(INFO, "Received from lookup " << from);
 
   {
+    // Process and update ds committee network info
     lock_guard<mutex> lock(m_mediator.m_mutexDSCommittee);
     for (const auto& dsguardupdate : vecOfDSGuardUpdateStruct) {
-      // Remove old ds guard IP info from blacklist exclude list
```
Comment (kaikawaliu) [inline src/libNode/Node.cpp]: Are these removals related to this PR?

File: src/libPersistence/Retriever.cpp [hunk 7]
```diff
@@ -50,26 +47,105 @@ bool Retriever::RetrieveTxBlocks(bool trimIncompletedBlocks) {
...
-    m_mediator.m_node->AddBlock(*block);
-  }
-
   /// Retrieve final block state delta from last DS epoch to
-  /// current TX epoch
+  /// current TX epoch and buffer the statedelta for each.
   for (const auto& block : blocks) {
     if (block->GetHeader().GetBlockNum() >= lastBlockNum + 1 - extra_txblocks) {
       bytes stateDelta;
       BlockStorage::GetBlockStorage().GetStateDelta(
```
Comment (kaikawaliu) [inline src/libPersistence/Retriever.cpp]: Just notice that, maybe we should handle the boolean return of this function.

File: src/libPersistence/Retriever.cpp [hunk 8]
```diff
@@ -50,26 +47,105 @@ bool Retriever::RetrieveTxBlocks(bool trimIncompletedBlocks) {
...
+  }
+
+  if ((lastBlockNum - extra_txblocks + 1) %
+          (INCRDB_DSNUMS_WITH_STATEDELTAS * NUM_FINAL_BLOCK_PER_POW) ==
+      0) {
+    // we must have latest state currently. so need not recreate states
+    LOG_GENERAL(INFO, "Current state is up-to-date until txblk :"
+                          << lastBlockNum - extra_txblocks);
+  } else {
+    // create states from last INCRDB_DSNUMS_WITH_STATEDELTAS *
```
Comment (kaikawaliu) [inline src/libPersistence/Retriever.cpp]: Hmmm sorry to say but actually this part of code is not needed. You can just keep DeserializeDelta from the last state checkpoint until the latest epoch.

File: src/libNode/Node.cpp (hunk 5 above)
Comment (sandipbhoir) [reply src/libNode/Node.cpp]: Haven't tried.. Just that we don't need the code below for new nodes joining.

File: src/libLookup/Lookup.cpp (hunk 1 above)
Comment (sandipbhoir) [reply src/libLookup/Lookup.cpp]: Fixed. Handled it differently.

File: src/libLookup/Lookup.cpp (hunk 4 above)
Comment (sandipbhoir) [reply src/libLookup/Lookup.cpp]: Since, we check on the count of deltas received against the request low/hignum and also verify the sender (lookup), additional check on state-delta hash can be considered for future improvements. If you agree.

File: src/libMessage/Messenger.cpp (hunk 3 above)
Comment (sandipbhoir) [reply src/libMessage/Messenger.cpp]: Fixed

File: src/libNode/Node.cpp (hunk 6 above)
Comment (sandipbhoir) [reply src/libNode/Node.cpp]: Fixed

File: src/libPersistence/Retriever.cpp (hunk 7 above)
Comment (sandipbhoir) [reply src/libPersistence/Retriever.cpp]: Fixed. If any of extra statedelta missing in BlockStorage, simple trim the extra tx blocks. we can then fetch the rest  from lookup.

File: src/libLookup/Lookup.cpp (hunk 4 above)
Comment (kaikawaliu) [reply src/libLookup/Lookup.cpp]: Yes, we can do that enhancement in the future

File: src/libNode/Node.cpp (hunk 6 above)
Comment (kaikawaliu) [reply src/libNode/Node.cpp]: Didn't find the code change. Do we really need to remove this part in this PR?

**Comment by kaikawaliu** [comment]
We can try the testing again with state trie repopulating enabled. To make it happen in every DS epoch. You can set `REPOPULATE_STATE_IN_DS` to 0, and `REPOPULATE_STATE_PER_N_DS` to 1

File: src/libLookup/Lookup.cpp [hunk 9]
```diff
@@ -1812,9 +1898,72 @@ bool Lookup::ProcessSetTxBlockFromSeed(const bytes& message,
...
+    return;
+  }
+
+  m_mediator.m_lookup->SendMessageToRandomSeedNode(getpowsubmission_message);
+}
+
 void Lookup::CommitTxBlocks(const vector<TxBlock>& txBlocks) {
   LOG_GENERAL(INFO, "[TxBlockVerif]"
                         << "Success");
+  uint64_t lowBlockNum = txBlocks.front().GetHeader().GetBlockNum();
```
Comment (KaustubhShamshery) [inline src/libLookup/Lookup.cpp]: unused?

File: src/libLookup/Lookup.cpp [hunk 10]
```diff
@@ -2142,7 +2142,8 @@ bool Lookup::ProcessSetStateDeltasFromSeed(const bytes& message,
         return false;
       }
       BlockStorage::GetBlockStorage().PutStateDelta(txBlkNum, delta);
-      if (txBlkNum + 1 % NUM_FINAL_BLOCK_PER_POW == 0) {
+      if (((txBlkNum + 1) % NUM_FINAL_BLOCK_PER_POW == 0) &&
+          (txBlkNum + NUM_FINAL_BLOCK_PER_POW < highBlockNum)) {
         if (!AccountStore::GetInstance().MoveUpdatesToDisk()) {
```
Comment (kaikawaliu) [inline src/libLookup/Lookup.cpp]: No param passing to MoveUpdatesToDisk means don't do repopulate. Aren't you going to repopulate at this point?

Comment (sandipbhoir) [reply src/libLookup/Lookup.cpp]: You are correct. This change is already made in PR#1540.
